src/libsyntax/parse/unescape.rs

```diff
@@ -66,6 +67,35 @@ where
     })
 }
 
+/// Takes a contents of a string literal (without quotes) and produces a
+/// sequence of characters or errors.
+/// NOTE: Raw strings do not perform any explicit character escaping, here we
+/// only translate CRLF to LF and produce errors on bare CR.
+pub(crate) fn unescape_raw_str<F>(literal_text: &str, mode: Mode, callback: &mut F)
+where
+    F: FnMut(Range<usize>, Result<char, EscapeError>),
+{
+    let mut byte_offset: usize = 0;
+
+    let mut chars = literal_text.chars().peekable();
+    while let Some(curr) = chars.next() {
+        let (result, scanned) = match (curr, chars.peek()) {
+            ('\r', Some('\n')) => {
+                chars.next();
+                (Ok('\n'), [Some('\r'), Some('\n')])
+            },
+            ('\r', _) =>
+                (Err(EscapeError::BareCarriageReturn), [Some('\r'), None]),
+            (c, _) if mode.is_bytes() && c > '\x7F' =>
+                (Err(EscapeError::NonAsciiCharInByteString), [Some(c), None]),
+            (c, _) => (Ok(c), [Some(c), None]),
+        };
+        let len_utf8: usize = scanned.iter().filter_map(|&x| x).map(char::len_utf8).sum();
```

Comment:
I think we don't need to maintain `scanned` separately, because it duplicates position info that is already present in the chars iterator. I suggest restoring this info from `chars.as_str().len()`

Reply:
I like how this version is more declarative (I find `initial_len`, `start`, `end` calculation dance [here](https://github.com/rust-lang/rust/blob/07c3967de9027735bfbf9c739f6a8ca14312c22d/src/libsyntax/parse/unescape.rs#L211-L243) not immediately obvious) but then you have to manually duplicate what you just scanned, which isn't ideal.

I can rewrite it using the start/end calculation if you'd like

Reply:
I slightly prefer to rewrite: I agree that it is spectacularly non-obvious, but it is guaranteed to be correct. in contrast, with the current approach one may accidentally bump twice, but return only one char in scanned.

Reply:
Done in https://github.com/rust-lang/rust/pull/60793/commits/63dc7da703759dd53536dd18a42ff65f39a2f9b4 - I think it's better this way, thanks!